### ISSUE TITLE
Default `www` is hardcoded and required

### ISSUE THREAD

**Issue by rlmestre** [mapiacompany/capacitor-codepush#33]
### Description

When building with Angular/Ionic/Capacitor for Android, the final assets output is located in `./android/app/src/main/assets/public`. The path used in the AppCenter CLI `updateContentsPath` flag, or `-c` is apparently relevant to the checks performed internally (or maybe used to name the resulting zip file.) Therefore, executing `appcenter codepush release -a "whatever" -c "./android/app/src/main/assets/public"` prevents the update files from being found. If you move the files to a `www` folder (or rename `public -> www`, it works. 

tl;dr `www` is hardcoded. If the resulting extracted dir is based on the zip filename, that should be used, or otherwise expose a config.

**Comment by mittermayr** [comment]
This was helpful to see, I was stuck there too — my deploy complained that it wasn't able to verify the signature (it said public key provided, but no JWT). Once I copied the `public` to the `www` folder, that step seems to have completed, but now it's complaining about JWT signature being invalid. 

Not sure if that's to do with the same www folder situation — please ignore my comment if that's mixing in a new problem...

`Provided key doesn't have a valid ASN.1 structure (first byte should be 0x30 == SEQUENCE`

Unzipping the codepush release returns a reasonably correct looking folder structure, including a `.codepushrelease` file (which it didn't before somehow). Public key is in Plist as `CodePushPublicKey`

**Comment by mittermayr** [comment]
Another update here (which may or may not happen for @rlmestre as well) — by copying the `public` to the `www` directory, everything seems to work, updates are detected, downloaded, verified and installed, except ...

Once the app restarts, it's back to the original `public` code again. I assume this is due to my poor fix of just copying `public` to `www` instead of this being perhaps fixed deeper in the code (probably what @rlmestre was hoping for).

So, how can one tell Ionic to use `www` instead of `public`? (Or, in Capacitor 3 style on iOS: `ios/App/App/www`

**Comment by Clovel** [comment]
We got to the point where Android Studio logs show 
```log
Sending plugin error: {"save":false,"callbackId":"55766507","pluginId":"CodePush","methodName":"preInstall","success":false,"error":{"message":"Could not get the package start page"}}
[CodePush] Preinstall failure. Could not get the package start page. StackTrace: Error: Could not get the package start page
```

I feel like this is directly related to the issue you are discussing. Not being able to find the start page could be caused by hardcoded paths. 

FYI, I have `./android/app/src/main/assets/public` as the path for `appcenter codepush release` commands too. The application runs fine, but when applying updates we get the above logs.

**Comment by Clovel** [comment]
Ok, so I fixed it by replacing the `"www/"` instances by `"public/"`. This is logical as Cordova used to generate `"www/"` directories (IIRC) whereas Capacitor generates `"public/"` directories.

I'll submit a MR later today.

**Comment by mittermayr** [comment]
So, @Clovel can you confirm the updates actually stick after app restarts? I thought I had fixed it on iOS by renaming `www/` to `public/` pretty much anywhere I could find it (manual edits, not bulk-replace) — this made the updates work (download and apply), but after app restarts, it simply featured the original version again.

**Comment by Clovel** [comment]
I confirmed it worked for Android. I'm currently battling with my Xcode installation to test this on my iPhone. I'll keep you posted.

**Comment by Clovel** [comment]
So, after testing on iOS, I get an exception when (or after ?) applying the update. Here are some of the logs I got before the exception.

## Logs
```log
⚡️  [log] - [CodePush] Installing update
⚡️  To Native ->  Filesystem stat 73249589
⚡️  To Native ->  Http request 73249590
ERROR MESSAGE:  {"errorMessage":"The file “unzipped” couldn’t be opened because there is no such file.","message":"The file “unzipped” couldn’t be opened because there is no such file."}
⚡️  [error] - {"errorMessage":"The file “unzipped” couldn’t be opened because there is no such file.","message":"The file “unzipped” couldn’t be opened because there is no such file."}
⚡️  To Native ->  Filesystem mkdir 73249591
⚡️  TO JS undefined
⚡️  To Native ->  Filesystem getUri 73249592
⚡️  TO JS {"uri":"file:\/\/\/var\/mobile\/Containers\/Data\/Application\/<applicationID>\/Documents\/codepush\/download\/unzipped\/"}
⚡️  To Native ->  CodePush unzip 73249593
⚡️  TO JS undefined
⚡️  To Native ->  Filesystem stat 73249594
ERROR MESSAGE:  {"errorMessage":"The file “hotcodepush.json” couldn’t be opened because there is no such file.","message":"The file “hotcodepush.json” couldn’t be opened because there is no such file."}
⚡️  [error] - {"errorMessage":"The file “hotcodepush.json” couldn’t be opened because there is no such file.","message":"The file “hotcodepush.json” couldn’t be opened because there is no such file."}
⚡️  To Native ->  Filesystem stat 73249595
ERROR MESSAGE:  {"errorMessage":"The file “versions” couldn’t be opened because there is no such file.","message":"The file “versions” couldn’t be opened because there is no such file."}
⚡️  [error] - {"errorMessage":"The file “versions” couldn’t be opened because there is no such file.","message":"The file “versions” couldn’t be opened because there is no such file."}
⚡️  To Native ->  Filesystem mkdir 73249596
⚡️  TO JS undefined
⚡️  To Native ->  Filesystem stat 73249597
ERROR MESSAGE:  {"errorMessage":"The file <versionHash> couldn’t be opened because there is no such file.","message":"The file <versionHash> couldn’t be opened because there is no such file."}
⚡️  [error] - {"errorMessage":"The file <versionHash> couldn’t be opened because there is no such file.","message":"The file <versionHash> couldn’t be opened because there is no such file."}
⚡️  To Native ->  Filesystem copy 73249598
⚡️  TO JS undefined
⚡️  [log] - [CodePush] Applying full update
⚡️  To Native ->  CodePush getPublicKey 73249599
2021-06-16 21:34:16.762908+0200 App[15993:5039757] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSPlaceholderDictionary initWithObjects:forKeys:count:]: attempt to insert nil object from objects[0]'
*** First throw call stack:
(0x195fae5b4 0x1aaa2d7a8 0x196020a80 0x19602cad8 0x195e9b360 0x195e8dc80 0x10090f160 0x100b2cd28 0x100b21c54 0x100e10064 0x100e11d90 0x100e19ea0 0x100e1acc4 0x100e27a00 0x1e19ca7a4 0x1e19d174c)
libc++abi: terminating with uncaught exception of type NSException
terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSPlaceholderDictionary initWithObjects:forKeys:count:]: attempt to insert nil object from objects[0]'
(lldb) 
```

**Comment by Clovel** [comment]
I strongly believe that this is caused by the `getPublicKey` call that returns `null` if no public key has been given. 

The plugin should be compatible with no signing. The android side works like that. 

@mittermayr perhaps you could try my fixes with a private/public key.

**Comment by Clovel** [comment]
I fixed this (https://github.com/mapiacompany/capacitor-codepush/issues/33#issuecomment-864928539) problem. MR coming soon. 

Thus, I can confirm that the CodePush updated do in fact stick.

**Comment by alexcroox** [comment]
@Clovel thank you so much for pushing this lib forward!

**Comment by lincolnthree** [comment]
Just for anyone else who finds this who already worked around the issue with a bit of scripting... 

You can get appcenter to bundle things correctly by creating a temporary folder with 'www' as the leaf node containing the built app/index.html and assets. 

This folder will now need to be  called 'public'.

`/temp-dir/www` -> `/temp-dir/public`
